src/mlpack/core/tree/vantage_point_tree/vantage_point_split_impl.hpp

```diff
@@ -0,0 +1,298 @@
+    // previous loop.
+    while ((!AssignToLeftSubtree(bound, data, vantagePoint, right, mu)) && (left <= right))
+      right--;
+  }
+
+  Log::Assert(left == right + 1);
+
+  return left;
+}
+
+template<typename BoundType, typename MatType>
+void VantagePointSplit<BoundType, MatType>::
+SelectVantagePoint(const BoundType& bound, const MatType& data,
+    const size_t begin, const size_t count, size_t& vantagePoint, ElemType& mu)
+{
+  arma::uvec vantagePointCandidates;
+
+  // Get no more than max(maxNumSamples, count) vantage point candidates
+  GetDistinctSamples(vantagePointCandidates, maxNumSamples, begin, count);
+
+  ElemType bestSpread = 0;
+
+  //  Evaluate eache candidate
+  for (size_t i = 0; i < vantagePointCandidates.n_rows; i++)
+  {
+    arma::uvec samples;
+
+    // Get no more than max(maxNumSamples, count) random samples
+    GetDistinctSamples(samples, maxNumSamples, begin, count);
+
+    // Calculate the second moment of the distance to the vantage point candidate
+    // using these random samples
+    const ElemType spread = GetSecondMoment(bound, data, samples,
+        vantagePointCandidates[i]);
+
+    if (spread > bestSpread)
+    {
+      bestSpread = spread;
+      vantagePoint = vantagePointCandidates[i];
+      //  Calculate the median value of the distance from the vantage point candidate
+      //  to these samples
+      GetMedian(bound, data, samples, vantagePoint, mu);
+    }    
+  }
+  assert(bestSpread > 0);
+}
+
+template<typename BoundType, typename MatType>
+void VantagePointSplit<BoundType, MatType>::
+GetDistinctSamples(arma::uvec& distinctSamples, const size_t numSamples,
+    const size_t begin, const size_t upperBound)
+{
+  if (upperBound > numSamples)
+  {
+    arma::Col<size_t> samples;
+
+    samples.zeros(upperBound);
+
+    for (size_t i = 0; i < numSamples; i++)
+      samples [ (size_t) math::RandInt(upperBound) ]++;
+
+    distinctSamples = arma::find(samples > 0);
+
+    distinctSamples += begin;
+  }
+  else
+  {
+    //  The node contains less points than requested
+    distinctSamples.set_size(upperBound);
+    for (size_t i = 0; i < upperBound; i++)
+      distinctSamples[i] = begin + i;
+  }
+}
+
+template<typename BoundType, typename MatType>
+void VantagePointSplit<BoundType, MatType>::
+GetMedian(const BoundType& bound, const MatType& data,
+    const arma::uvec& samples, const size_t vantagePoint, ElemType& mu)
+{
+  std::vector<SortStruct<ElemType>> sorted(samples.n_rows);
+
+  for (size_t i = 0; i < samples.n_rows; i++)
+  {
+    sorted[i].point = samples[i];
+    sorted[i].dist = bound.Metric().Evaluate(data.col(vantagePoint),
+        data.col(samples[i]));
+  }
+
+  //  Sort samples according to the distance to the vantage point
+  std::sort(sorted.begin(), sorted.end(), StructComp<ElemType>);
+
+  //  Get the midian value
+  mu = bound.Metric().Evaluate(data.col(vantagePoint),
+      data.col(sorted[sorted.size() / 2].point));
```

Comment:
Could you instead use Armadillo's `median()` function here?  You don't need the index of the median, just the median value itself, so I think Armadillo could work fine, and I think it would be less code (and possibly faster too).